brave/brave-core
# fixed crash with search extension

fix: https://github.com/brave/brave-browser/issues/15224
fix: https://github.com/brave/brave-browser/issues/10601
fix: https://github.com/brave/brave-browser/issues/14218

With this PR, brave uses search provider from extension by default on all platform like chromium does.

Crash comes from absense of exension search provider data in private profile's TemplateURLService because
we are using different TemplateURLService for private profile. When search provider extension is loaded/installed, browser doesn't notify about this adding to private profile's TUS.
To fix this, each SearchEngineProviderService sets extension's template url data to its service when normal profile's search provider comes from extension.

## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:
`npm run test brave_browser_tests -- --filter=*ExtensionSearchProviderWithPrivateWindow*`

1. Launch browser
2. Install ecosia extension - https://chrome.google.com/webstore/detail/ecosia-the-search-engine/eedlgdlajadkbbjoobobefphmfkcchfk/related
3. Check Normal window's default search provider is ecosia
4. Open private window and check ecosia is default search provider
5. Turn ddg button on doesn't change search provider - still using ecosia
6. Turn ddg button off doesn't change search provider - still using ecosia
7. Goto ecosia extension details and toggle `Allow in private` repeatedly
8. Check changing `Allow in privatge` option doesn't affect private window's search provider - still using ecosia
9. Disable ecosia extension and check normal and private window uses default search provider
10. Toggle DDG button on in private NTP and check search provider is changed to DDG
